File: frontends/admin-ui/src/types/revenue.ts
export type MoneyAmount = {
  amount: number;
  currency: string;
};

export type RevenuePlanMixItem = {
  plan: string;
  orgs: number;
  mrr: number | null;
};

export type RevenueAddonMixItem = {
  addon: string;
  orgs: number;
  mrr: number;
};

export type RevenueSummaryResponse = {
  as_of: string;
  mrr: MoneyAmount;
  arr: MoneyAmount;
  active_orgs: number;
  overdue_orgs: number;
  overdue_amount: number;
  usage_revenue_mtd: number;
  plan_mix: RevenuePlanMixItem[];
  addon_mix: RevenueAddonMixItem[];
  overdue_buckets: RevenueOverdueBucket[];
};

export type RevenueOverdueBucket = {
  bucket: string;
  label: string;
  orgs: number;
  amount: number;
};

export type RevenueOverdueItem = {
  org_id: number;
  org_name: string | null;
  invoice_id: string;
  due_at: string | null;
  overdue_days: number;
  amount: number;
  currency: string | null;
  subscription_plan: string | null;
  subscription_status: string | null;
};

export type RevenueOverdueResponse = {
  items: RevenueOverdueItem[];
  total: number;
  limit: number;
  offset: number;
};

File: platform/processing-core/app/schemas/admin/revenue.py
from __future__ import annotations

from datetime import date, datetime
from decimal import Decimal

from pydantic import BaseModel


class MoneyAmount(BaseModel):
    amount: Decimal
    currency: str


class RevenuePlanMixItem(BaseModel):
    plan: str
    orgs: int
    mrr: Decimal | None = None


class RevenueAddonMixItem(BaseModel):
    addon: str
    orgs: int
    mrr: Decimal


class RevenueSummaryResponse(BaseModel):
    as_of: date
    mrr: MoneyAmount
    arr: MoneyAmount
    active_orgs: int
    overdue_orgs: int
    overdue_amount: Decimal
    usage_revenue_mtd: Decimal
    plan_mix: list[RevenuePlanMixItem]
    addon_mix: list[RevenueAddonMixItem]
    overdue_buckets: list["RevenueOverdueBucket"]


class RevenueOverdueBucket(BaseModel):
    bucket: str
    label: str
    orgs: int
    amount: Decimal


class RevenueOverdueItem(BaseModel):
    org_id: int
    org_name: str | None = None
    invoice_id: str
    due_at: datetime | None = None
    overdue_days: int
    amount: Decimal
    currency: str | None = None
    subscription_plan: str | None = None
    subscription_status: str | None = None


class RevenueOverdueResponse(BaseModel):
    items: list[RevenueOverdueItem]
    total: int
    limit: int
    offset: int


class RevenueUsageMeterItem(BaseModel):
    ref_code: str | None = None
    quantity: Decimal
    amount: Decimal


class RevenueUsageResponse(BaseModel):
    period_from: date
    period_to: date
    meters: list[RevenueUsageMeterItem]
